File: src/domain/overloadEngine.ts
import { Exercise, WorkoutSession, WorkoutSet, EngineResult, BannerType } from '../types';

export class ArgumentError extends Error {
  constructor(message: string) {
    super(`ArgumentError (OCL Invariant Violation): ${message}`);
    this.name = 'ArgumentError';
  }
}

export class StateError extends Error {
  constructor(message: string) {
    super(`StateError (Pre-condition Failure): ${message}`);
    this.name = 'StateError';
  }
}

export function validateExerciseOCL(exercise: Exercise): void {
  if (exercise.sets <= 0) {
    throw new ArgumentError(`sets must be > 0 (got ${exercise.sets})`);
  }
  if (exercise.baseIncrement <= 0) {
    throw new ArgumentError(`baseIncrement must be > 0 (got ${exercise.baseIncrement})`);
  }
  if (exercise.maxWeightCapacity <= 0) {
    throw new ArgumentError(`maxWeightCapacity must be > 0 (got ${exercise.maxWeightCapacity})`);
  }
  if (exercise.baseIncrement > exercise.maxWeightCapacity) {
    throw new ArgumentError(
      `baseIncrement (${exercise.baseIncrement}) cannot exceed maxWeightCapacity (${exercise.maxWeightCapacity})`
    );
  }
}

export function validateWorkoutSetOCL(set: WorkoutSet): void {
  if (set.setNumber <= 0) {
    throw new ArgumentError(`setNumber must be > 0 (got ${set.setNumber})`);
  }
  if (set.targetWeight <= 0) {
    throw new ArgumentError(`targetWeight must be > 0 (got ${set.targetWeight})`);
  }
  if (set.targetReps <= 0) {
    throw new ArgumentError(`targetReps must be > 0 (got ${set.targetReps})`);
  }
  if (set.actualWeight < 0) {
    throw new ArgumentError(`actualWeight must be >= 0 (got ${set.actualWeight})`);
  }
  if (set.actualReps < 0) {
    throw new ArgumentError(`actualReps must be >= 0 (got ${set.actualReps})`);
  }
}

export class OverloadEngine {
  constructor(private exercise: Exercise) {
    validateExerciseOCL(exercise);
  }

  public computeNextTarget(
    current: WorkoutSession,
    historyForExercise: WorkoutSession[] = []
  ): EngineResult {
    // 1. Validate Pre-conditions
    if (!current.sets || current.sets.length === 0) {
      throw new StateError('Workout session contains no sets.');
    }

    for (const set of current.sets) {
      validateWorkoutSetOCL(set);
      if (!set.isCompleted) {
        throw new StateError(`Set #${set.setNumber} is incomplete. Check off all sets before evaluating.`);
      }
    }

    // 2. Anti-contamination rule: verify historical sessions match exercise ID
    for (const h of historyForExercise) {
      if (h.exerciseId !== this.exercise.id) {
        throw new ArgumentError(
          `Historical session ${h.id} belongs to exercise ${h.exerciseId}, not ${this.exercise.id}`
        );
      }
    }

    // Determine current target weight & reps
    const currentTargetWeight = current.sets[0].actualWeight > 0 ? current.sets[0].actualWeight : current.sets[0].targetWeight;
    const currentTargetReps = current.sets[0].targetReps;

    // Check if current session passed target reps across ALL sets
    const currentPassed = current.sets.every((s) => s.actualReps >= s.targetReps);

    if (currentPassed) {
      const nextTargetWeight = Math.min(
        currentTargetWeight + this.exercise.baseIncrement,
        this.exercise.maxWeightCapacity
      );

      const isCapReached = nextTargetWeight === this.exercise.maxWeightCapacity && currentTargetWeight < this.exercise.maxWeightCapacity;

      return {
        nextTargetWeight,
        nextTargetReps: currentTargetReps,
        isProgression: true,
        isDeload: false,
        message: isCapReached
          ? `Max Capacity Reached! Target set to ${nextTargetWeight} lbs (capped at max ${this.exercise.maxWeightCapacity} lbs).`
          : `Progressive Overload Achieved! Next Target: ${nextTargetWeight} lbs × ${currentTargetReps} reps (+${this.exercise.baseIncrement} lbs).`,
        bannerType: BannerType.SUCCESS,
      };
    }

    // Check 3-Strike Deload Penalty across 3 consecutive sessions
    const allSessions = [...historyForExercise, current].sort(
      (a, b) => new Date(a.date).getTime() - new Date(b.date).getTime()
    );

    const recentSessions = allSessions.length >= 3 ? allSessions.slice(-3) : allSessions;

    const isThreeStrikeDeload =
      recentSessions.length === 3 &&
      recentSessions.every((session) =>
        session.sets.some((s) => s.actualReps < s.targetReps)
      );

    if (isThreeStrikeDeload) {
      let rawDeload = currentTargetWeight * 0.9;
      // Round to nearest 0.5 lbs for practical gym weight plates
      const nextTargetWeight = Math.round(rawDeload * 2) / 2;

      return {
        nextTargetWeight,
        nextTargetReps: currentTargetReps,
        isProgression: false,
        isDeload: true,
        message: `3-Strike Deload Penalty Triggered! Target reduced by 10% to ${nextTargetWeight} lbs. Focus on form and recovery.`,
        bannerType: BannerType.ERROR,
      };
    }

    const failedCount = recentSessions.filter((s) =>
      s.sets.some((set) => set.actualReps < set.targetReps)
    ).length;

    return {
      nextTargetWeight: currentTargetWeight,
      nextTargetReps: currentTargetReps,
      isProgression: false,
      isDeload: false,
      message: `Target Maintained at ${currentTargetWeight} lbs. (Strike ${failedCount}/3 before automatic deload penalty).`,
      bannerType: BannerType.INFO,
    };
  }
}

File: lib/domain/overload_engine.dart
import 'dart:math';
import 'exercise.dart';
import 'workout_session.dart';
import 'workout_set.dart';

class OverloadEngine {
  final Exercise exercise;

  OverloadEngine({required this.exercise});

  /// Computes the next target set based on current and historical workout sessions.
  /// Throws [StateError] if pre-conditions are violated.
  /// Throws [ArgumentError] if OCL invariants are breached.
  WorkoutSet computeNextTarget({
    required WorkoutSession current,
    List<WorkoutSession> historyForExercise = const [],
  }) {
    // 1. Pre-condition checks
    if (current.sets.isEmpty) {
      throw StateError('Pre-condition Failure: Workout session contains no sets.');
    }

    for (final set in current.sets) {
      if (set.actualWeight < 0 || set.actualReps < 0) {
        throw StateError(
          'Pre-condition Failure: Actual weight (${set.actualWeight}) and reps (${set.actualReps}) cannot be negative.',
        );
      }
      if (!set.isCompleted) {
        throw StateError('Pre-condition Failure: Set #${set.setNumber} is incomplete.');
      }
    }

    // Anti-contamination verification: ensure all historical sessions match this exercise
    for (final h in historyForExercise) {
      if (h.exerciseId != exercise.id) {
        throw ArgumentError(
          'State Isolation Violation: Historical session ${h.id} belongs to exercise ${h.exerciseId}, not ${exercise.id}.',
        );
      }
    }

    // Determine if current session passed
    final currentPassed = current.sets.every((s) => s.actualReps >= s.targetReps);

    // Calculate current baseline target weight
    final currentTargetWeight = current.sets.first.targetWeight;
    final currentTargetReps = current.sets.first.targetReps;

    double nextTargetWeight;

    if (currentPassed) {
      // Rule 1: Success -> Increment weight up to max capacity
      nextTargetWeight = min(
        currentTargetWeight + exercise.baseIncrement,
        exercise.maxWeightCapacity,
      );
    } else {
      // Check 3-strike deload penalty rule across 3 consecutive sessions
      // Combine history with current session (sorted chronologically)
      final allSessions = [...historyForExercise, current]..sort((a, b) => a.date.compareTo(b.date));

      // Examine up to the last 3 sessions for this exercise
      final recentSessions = allSessions.length >= 3
          ? allSessions.sublist(allSessions.length - 3)
          : allSessions;

      final isThreeStrikeDeload = recentSessions.length == 3 &&
          recentSessions.every(
            (session) => session.sets.any((s) => s.actualReps < s.targetReps),
          );

      if (isThreeStrikeDeload) {
        // Rule 2: 3-Strike Deload -> Deload by 10%
        nextTargetWeight = currentTargetWeight * 0.9;
        // Round to nearest 0.5 or 2.5 lbs for practical gym use
        nextTargetWeight = (nextTargetWeight * 2).roundToDouble() / 2;
      } else {
        // Maintenance: keep same weight
        nextTargetWeight = currentTargetWeight;
      }
    }

    return WorkoutSet(
      id: 'next_${DateTime.now().millisecondsSinceEpoch}',
      setNumber: 1,
      targetWeight: nextTargetWeight,
      targetReps: currentTargetReps,
      actualWeight: 0.0,
      actualReps: 0,
      isCompleted: false,
    );
  }
}
